File: src/libsokoengine/game/octoban_tessellation.cpp
#include "octoban_tessellation.hpp"

#include "pusher_step.hpp"
#include "snapshot.hpp"

using sokoengine::io::CellOrientation;
using std::invalid_argument;
using std::string;

namespace sokoengine {
namespace game {

const Directions &OctobanTessellation::legal_directions() const {
  static const Directions retv = {
    Direction::LEFT,
    Direction::RIGHT,
    Direction::UP,
    Direction::DOWN,
    Direction::NORTH_EAST,
    Direction::NORTH_WEST,
    Direction::SOUTH_EAST,
    Direction::SOUTH_WEST};
  return retv;
}

position_t OctobanTessellation::neighbor_position(
  position_t       position,
  const Direction &direction,
  board_size_t     width,
  board_size_t     height
) const {
  position_t row = index_y(position, width), column = index_x(position, width);

  // clang-format off
  if (
    cell_orientation(position, width, height) != CellOrientation::OCTAGON
    && (
      direction == Direction::NORTH_EAST
      || direction == Direction::NORTH_WEST
      || direction == Direction::SOUTH_EAST
      || direction == Direction::SOUTH_WEST
    )
  ) {
    return Config::NO_POS;
  }
  // clang-format off

  switch (direction) {
    case Direction::LEFT:
      column += -1;
      break;
    case Direction::UP:
      row += -1;
      break;
    case Direction::RIGHT:
      column += 1;
      break;
    case Direction::DOWN:
      row += 1;
      break;
    case Direction::NORTH_WEST:
      row += -1;
      column += -1;
      break;
    case Direction::NORTH_EAST:
      row += -1;
      column += 1;
      break;
    case Direction::SOUTH_WEST:
      row += 1;
      column += -1;
      break;
    case Direction::SOUTH_EAST:
      row += 1;
      column += 1;
      break;
    default:
      throw invalid_argument(
        "Unsupported direction " + implementation::direction_str(direction)
        + " for OctobanTessellation!"
      );
  }
  if (is_on_board_2d(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::NO_POS;
}

PusherStep OctobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
    case io::Snapshot::l:
      return PusherStep(Direction::LEFT, Config::NO_ID);
    case io::Snapshot::L:
      return PusherStep(Direction::LEFT, Config::DEFAULT_ID);

    case io::Snapshot::u:
      return PusherStep(Direction::UP, Config::NO_ID);
    case io::Snapshot::U:
      return PusherStep(Direction::UP, Config::DEFAULT_ID);

    case io::Snapshot::r:
      return PusherStep(Direction::RIGHT, Config::NO_ID);
    case io::Snapshot::R:
      return PusherStep(Direction::RIGHT, Config::DEFAULT_ID);

    case io::Snapshot::d:
      return PusherStep(Direction::DOWN, Config::NO_ID);
    case io::Snapshot::D:
      return PusherStep(Direction::DOWN, Config::DEFAULT_ID);

    case io::Snapshot::w:
      return PusherStep(Direction::NORTH_WEST, Config::NO_ID);
    case io::Snapshot::W:
      return PusherStep(Direction::NORTH_WEST, Config::DEFAULT_ID);

    case io::Snapshot::e:
      return PusherStep(Direction::SOUTH_EAST, Config::NO_ID);
    case io::Snapshot::E:
      return PusherStep(Direction::SOUTH_EAST, Config::DEFAULT_ID);

    case io::Snapshot::n:
      return PusherStep(Direction::NORTH_EAST, Config::NO_ID);
    case io::Snapshot::N:
      return PusherStep(Direction::NORTH_EAST, Config::DEFAULT_ID);

    case io::Snapshot::s:
      return PusherStep(Direction::SOUTH_WEST, Config::NO_ID);
    case io::Snapshot::S:
      return PusherStep(Direction::SOUTH_WEST, Config::DEFAULT_ID);

    default:
      throw invalid_argument(
        string("Illegal PusherStep character '") + rv + "' in OctobanTessellation!"
      );
  }
}

char OctobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
    case Direction::LEFT:
      return rv.is_push_or_pull() ? io::Snapshot::L : io::Snapshot::l;
    case Direction::UP:
      return rv.is_push_or_pull() ? io::Snapshot::U : io::Snapshot::u;
    case Direction::RIGHT:
      return rv.is_push_or_pull() ? io::Snapshot::R : io::Snapshot::r;
    case Direction::DOWN:
      return rv.is_push_or_pull() ? io::Snapshot::D : io::Snapshot::d;
    case Direction::NORTH_WEST:
      return rv.is_push_or_pull() ? io::Snapshot::W : io::Snapshot::w;
    case Direction::SOUTH_EAST:
      return rv.is_push_or_pull() ? io::Snapshot::E : io::Snapshot::e;
    case Direction::NORTH_EAST:
      return rv.is_push_or_pull() ? io::Snapshot::N : io::Snapshot::n;
    case Direction::SOUTH_WEST:
      return rv.is_push_or_pull() ? io::Snapshot::S : io::Snapshot::s;
    default:
      throw invalid_argument(
        "Illegal PusherStep direction " + implementation::direction_str(rv.direction())
        + " in OctobanTessellation!"
      );
  }
}

CellOrientation OctobanTessellation::cell_orientation(
  position_t pos, board_size_t width, board_size_t height
) const {
  position_t column = index_column(pos, width);
  position_t row    = index_row(pos, width);
  return ((column + (row % 2)) % 2 == 0) ? CellOrientation::OCTAGON
                                         : CellOrientation::DEFAULT;
}

} // namespace game
} // namespace sokoengine

File: src/sokoenginepy/game/octoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, Optional, Tuple

from ..io import CellOrientation, Snapshot
from .base_tessellation import COLUMN, ROW, BaseTessellation, index_1d, is_on_board_2d
from .config import Direction
from .utilities import inverted


class OctobanTessellation(BaseTessellation):
    """
    Board space is laid out on alternating squares and octagons with origin of
    coordinate system being octagon. Tessellation allows all 8 directions of movement
    from Direction and depending on current pusher position some of these directions do
    not result in successful move.

    Direction <-> character mapping:

    ====  ==========  =====  ==========  ====  ==========  ====  ==========
    UP    NORTH_EAST  RIGHT  SOUTH_EAST  DOWN  SOUTH_WEST  LEFT  NORTH_WEST
    ====  ==========  =====  ==========  ====  ==========  ====  ==========
    u, U  n, N        r, R   e, E        d, D  s, S        l, L  w, W
    ====  ==========  =====  ==========  ====  ==========  ====  ==========
    """

    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.UP,
        Direction.DOWN,
        Direction.NORTH_EAST,
        Direction.NORTH_WEST,
        Direction.SOUTH_EAST,
        Direction.SOUTH_WEST,
    )

    _CHR_TO_PUSHER_STEP: Final[Dict[str, Tuple[Direction, bool]]] = {
        Snapshot.l: (Direction.LEFT, False),
        Snapshot.L: (Direction.LEFT, True),
        Snapshot.r: (Direction.RIGHT, False),
        Snapshot.R: (Direction.RIGHT, True),
        Snapshot.u: (Direction.UP, False),
        Snapshot.U: (Direction.UP, True),
        Snapshot.d: (Direction.DOWN, False),
        Snapshot.D: (Direction.DOWN, True),
        Snapshot.w: (Direction.NORTH_WEST, False),
        Snapshot.W: (Direction.NORTH_WEST, True),
        Snapshot.e: (Direction.SOUTH_EAST, False),
        Snapshot.E: (Direction.SOUTH_EAST, True),
        Snapshot.n: (Direction.NORTH_EAST, False),
        Snapshot.N: (Direction.NORTH_EAST, True),
        Snapshot.s: (Direction.SOUTH_WEST, False),
        Snapshot.S: (Direction.SOUTH_WEST, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Dict[Tuple[Direction, bool], str]] = inverted(
        _CHR_TO_PUSHER_STEP
    )

    _NEIGHBOR_SHIFT: Final[Dict[Direction, Tuple[int, int]]] = {
        Direction.LEFT: (0, -1),
        Direction.RIGHT: (0, 1),
        Direction.UP: (-1, 0),
        Direction.DOWN: (1, 0),
        Direction.NORTH_WEST: (-1, -1),
        Direction.NORTH_EAST: (-1, 1),
        Direction.SOUTH_WEST: (1, -1),
        Direction.SOUTH_EAST: (1, 1),
    }

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> Optional[int]:
        if self.cell_orientation(
            position, board_width, board_height
        ) != CellOrientation.OCTAGON and (
            direction == Direction.NORTH_EAST
            or direction == Direction.NORTH_WEST
            or direction == Direction.SOUTH_EAST
            or direction == Direction.SOUTH_WEST
        ):
            return None

        row = ROW(position, board_width)
        column = COLUMN(position, board_width)
        row_shift, column_shift = self._NEIGHBOR_SHIFT.get(direction, (None, None))

        if row_shift is None or column_shift is None:
            raise ValueError(direction)

        row += row_shift
        column += column_shift

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return None

    def cell_orientation(
        self, position: int, board_width: int, board_height: int
    ) -> CellOrientation:
        row = ROW(position, board_width)
        column = COLUMN(position, board_width)
        return (
            CellOrientation.OCTAGON
            if (column + (row % 2)) % 2 == 0
            else CellOrientation.DEFAULT
        )
